File: scripts/memory-deep-verify/Deep.cpp
// Deep.cpp — iterator invalidation, the cache hierarchy, stack vs heap, hashing.
// Measured, not asserted. g++ -std=c++17 -O2 -o deep Deep.cpp && ./deep
#include <algorithm>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <numeric>
#include <random>
#include <set>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

int main() {
    cout << boolalpha;

    /* ---------------- D01  iterator invalidation ---------------- */
    {
        vector<int> v{1, 2, 3};
        v.reserve(3);
        int* before = v.data();
        size_t capBefore = v.capacity();
        v.push_back(4);                       // forces a reallocation
        int* after = v.data();
        cout << "D01 vector: capacity " << capBefore << " -> " << v.capacity()
             << ", buffer moved -> " << (before != after)
             << "  EVERY iterator, pointer and reference into it is now dangling\n";

        vector<int> w{1, 2, 3};
        w.reserve(100);                        // room to spare
        int* p0 = w.data();
        w.push_back(4);                        // no reallocation this time
        cout << "D01safe with reserve(100) headroom, push_back did NOT move the buffer -> "
             << (p0 == w.data()) << "  (reserve is how you keep pointers valid)\n";
    }
    {
        // erase invalidates from the erase point onward; the RETURN VALUE is the fix
        vector<int> v{1, 2, 3, 4, 5};
        for (auto it = v.begin(); it != v.end();)
            if (*it % 2 == 0) it = v.erase(it); else ++it;
        ostringstream o; for (int x : v) o << x << " ";
        cout << "D01erase erase() returns the next valid iterator -> " << o.str() << "\n";
    }
    {
        // node-based containers do NOT invalidate other elements
        list<int> l{1, 2, 3};
        auto it = next(l.begin());             // -> 2
        l.push_front(0);
        l.push_back(9);
        cout << "D01node std::list: after inserting at both ends, the old iterator still reads "
             << *it << "  — node containers keep every other iterator valid\n";

        map<int, int> m{{1, 10}, {2, 20}};
        auto mit = m.find(2);
        m[3] = 30;
        cout << "D01map std::map: after inserting a new key, the old iterator still reads "
             << mit->second << "  (same guarantee)\n";

        unordered_map<int, int> um;
        um.reserve(2);
        for (int i = 0; i < 2; ++i) um[i] = i;
        float lf0 = um.load_factor();
        size_t b0 = um.bucket_count();
        for (int i = 2; i < 50; ++i) um[i] = i;   // forces rehash
        cout << "D01hash unordered_map: buckets " << b0 << " -> " << um.bucket_count()
             << ", load_factor " << lf0 << " -> " << um.load_factor()
             << "  — a REHASH invalidates iterators, but NOT references to values\n";
    }

    /* ---------------- D02  the cache hierarchy ---------------- */
    {
        /* Walk a fixed number of elements with a growing stride. While the working set
         * fits a cache level the time per access is flat; each time it spills to the
         * next level the time steps up. The steps ARE the hierarchy. */
        const size_t N = 1 << 22;              // 4M ints = 16 MB, bigger than most L3
        vector<int> a(N, 1);
        volatile long long sink = 0;

        auto timeStride = [&](size_t stride) {
            const size_t touches = 1 << 20;    // the same COUNT of reads every time
            auto t0 = chrono::steady_clock::now();
            long long s2 = 0;
            for (size_t rep = 0; rep < 4; ++rep)
                for (size_t i = 0, k = 0; k < touches; ++k, i += stride)
                    s2 += a[i & (N - 1)];
            auto t1 = chrono::steady_clock::now();
            sink += s2;
            return chrono::duration<double, milli>(t1 - t0).count();
        };
        // best of 3 on each, so one scheduling hiccup cannot flip a verdict
        auto best = [&](size_t stride) {
            double b = 1e18;
            for (int r = 0; r < 3; ++r) b = min(b, timeStride(stride));
            return b;
        };
        double s1 = best(1), s16 = best(16), s1024 = best(1024);
        cout << "D02 the SAME number of reads, walked with a growing stride:\n"
             << "D02step stride 16 slower than stride 1 -> " << (s16 > s1)
             << "   stride 1024 slower than stride 16 -> " << (s1024 > s16)
             << "   1024 slower than 1 -> " << (s1024 > s1) << "\n";
        cout << "D02why identical instruction count, identical O(n). The only thing that changed\n"
             << "        is how many CACHE LINES each pass has to fetch.\n";
        cout << "D02line a line is 64 bytes = 16 ints. At stride 1 you use all 16 ints of every\n"
             << "        line you pay for; at stride 16 and beyond you use ONE, so you pay 16x\n"
             << "        the memory traffic for the same answer — and the prefetcher stops helping.\n";
    }
    {
        // row-major vs column-major on the same matrix
        const int M = 1024;
        vector<int> g(size_t(M) * M, 1);
        volatile long long sink = 0;
        auto t0 = chrono::steady_clock::now();
        long long r = 0;
        for (int i = 0; i < M; ++i) for (int j = 0; j < M; ++j) r += g[size_t(i) * M + j];
        auto t1 = chrono::steady_clock::now();
        long long c = 0;
        for (int j = 0; j < M; ++j) for (int i = 0; i < M; ++i) c += g[size_t(i) * M + j];
        auto t2 = chrono::steady_clock::now();
        sink += r + c;
        double rowMs = chrono::duration<double, milli>(t1 - t0).count();
        double colMs = chrono::duration<double, milli>(t2 - t1).count();
        cout << "D02order 1024x1024 sum, same total: " << (r == c)
             << ".  column-major slower than row-major -> " << (colMs > rowMs)
             << "  — swapping two loop lines, nothing else\n";
    }

    /* ---------------- D03  stack vs heap ---------------- */
    {
        int local = 0;
        vector<int> heapv(4);
        cout << "D03 a local int lives on the STACK; new/malloc and every vector buffer live on\n"
             << "        the HEAP; the vector OBJECT itself sits wherever you declared it.\n";
        cout << "D03size sizeof(vector<int>) object=" << sizeof(heapv)
             << "B on the stack, holding a pointer to " << heapv.capacity() * sizeof(int)
             << "B on the heap. local=" << local << "\n";
        cout << "D03cost the stack is a pointer bump — allocation is one instruction and freeing is\n"
             << "        automatic at scope exit. The heap needs a real allocator and a matching\n"
             << "        free, which is why a hot loop that news per iteration is slow.\n";
    }
    {
        /* Measure the FRAME SIZE rather than recursing until the stack dies: a real
         * stack overflow is a crash the process cannot catch, and crashing the
         * verification run to prove a point about crashing would be a poor trade. */
        struct R {
            static ptrdiff_t probe(int d, const char* prev) {
                char here;
                if (d == 0) return probe(1, &here);
                return prev - &here;          // frames grow downward on x86-64
            }
        };
        char anchor;
        ptrdiff_t frame = R::probe(0, &anchor);
        if (frame < 0) frame = -frame;
        cout << "D03frame one recursive frame of this shape measures " << frame << " bytes\n";
        cout << "D03depth at that size, an 8 MB Linux stack allows about "
             << (8 * 1024 * 1024 / (frame ? frame : 1))
             << " levels; a 1 MB Windows stack allows about "
             << (1 * 1024 * 1024 / (frame ? frame : 1)) << "\n";
        cout << "D03risk a DFS on 1e5 nodes with a few locals per frame is genuinely close to the\n"
             << "        Windows limit, and overflowing the stack is a CRASH, not an exception you\n"
             << "        can catch. Convert to an explicit stack when depth can reach 1e5.\n";
    }

    /* ---------------- D04  hash containers ---------------- */
    {
        unordered_map<int, int> m;
        ostringstream growth;
        size_t lastB = 0;
        for (int i = 0; i < 1000; ++i) {
            m[i] = i;
            if (m.bucket_count() != lastB) { lastB = m.bucket_count(); growth << lastB << " "; }
        }
        cout << "D04 bucket counts as it grew: " << growth.str() << "\n";
        cout << "D04lf max_load_factor=" << m.max_load_factor()
             << " current=" << m.load_factor()
             << " buckets=" << m.bucket_count() << " for " << m.size() << " entries\n";
        cout << "D04mem libstdc++ unordered_map is buckets + a NODE per entry (~"
             << (sizeof(void*) * 2 + sizeof(pair<const int, int>))
             << "B each), so it is roughly 3-4x a vector holding the same pairs\n";
    }
    {
        // collisions: all keys landing in one bucket turns O(1) into O(n)
        unordered_map<size_t, int> m;
        m.reserve(1024);
        size_t B = m.bucket_count();
        for (int i = 0; i < 200; ++i) m[size_t(i) * B] = i;   // every key hashes to bucket 0
        size_t worst = 0;
        for (size_t b = 0; b < m.bucket_count(); ++b) worst = max(worst, m.bucket_size(b));
        unordered_map<size_t, int> fair;
        fair.reserve(1024);
        for (int i = 0; i < 200; ++i) fair[i] = i;
        size_t fairWorst = 0;
        for (size_t b = 0; b < fair.bucket_count(); ++b) fairWorst = max(fairWorst, fair.bucket_size(b));
        cout << "D04coll 200 keys chosen as multiples of bucket_count -> worst bucket holds "
             << worst << " entries; ordinary keys -> worst bucket holds " << fairWorst << "\n";
        cout << "D04why std::hash<size_t> is the IDENTITY on libstdc++, so an adversary who knows\n"
             << "        your bucket count can force every key into one chain. That is a real\n"
             << "        anti-hash attack in competitive programming: mix the key first.\n";
    }
    {
        // ordered vs unordered: the tradeoff, timed
        const int N = 200000;
        vector<int> keys(N);
        iota(keys.begin(), keys.end(), 0);
        mt19937 rng(42);
        shuffle(keys.begin(), keys.end(), rng);

        unordered_map<int, int> um; um.reserve(N * 2);
        map<int, int> om;
        auto t0 = chrono::steady_clock::now();
        for (int k : keys) um[k] = k;
        auto t1 = chrono::steady_clock::now();
        for (int k : keys) om[k] = k;
        auto t2 = chrono::steady_clock::now();
        volatile long long s = 0;
        for (int k : keys) s += um[k];
        auto t3 = chrono::steady_clock::now();
        for (int k : keys) s += om[k];
        auto t4 = chrono::steady_clock::now();

        double uIns = chrono::duration<double, milli>(t1 - t0).count();
        double oIns = chrono::duration<double, milli>(t2 - t1).count();
        double uLook = chrono::duration<double, milli>(t3 - t2).count();
        double oLook = chrono::duration<double, milli>(t4 - t3).count();
        cout << "D04cmp " << N << " random keys: unordered_map faster to build -> " << (uIns < oIns)
             << ", faster to look up -> " << (uLook < oLook)
             << "  — but map gives you sorted iteration and lower_bound, which no hash can\n";
    }

    return 0;
}

File: scripts/memory-deep-verify/Deep.java
// Deep.java — iterator invalidation, the cache hierarchy, stack vs heap, hashing.
// Measured, not asserted. java Deep.java  (JDK 11 single-file source launcher)
import java.util.*;

public class Deep {

    /* D03: measure the frame size by recursing exactly twice and diffing the depth
     * the JVM reports, rather than recursing until StackOverflowError. Catching SOE
     * is legal in Java but leaves the JVM in a state you should not keep working in,
     * and a verification run should not deliberately damage its own process. */
    static int probeDepth(int d, int cap) {
        if (d >= cap) return d;
        return probeDepth(d + 1, cap);
    }

    public static void main(String[] args) {

        /* ---------------- D01  iterator invalidation ---------------- */
        {
            List<Integer> v = new ArrayList<>(List.of(1, 2, 3, 4, 5));
            String cme;
            try {
                for (Integer x : v) if (x % 2 == 0) v.remove(x);
                cme = "no throw";
            } catch (ConcurrentModificationException e) {
                cme = "ConcurrentModificationException";
            }
            System.out.println("D01 for-each + structural modification -> " + cme
                + "   Java DETECTS it with a modCount check and fails fast");

            List<Integer> w = new ArrayList<>(List.of(1, 2, 3, 4, 5));
            Iterator<Integer> it = w.iterator();
            while (it.hasNext()) if (it.next() % 2 == 0) it.remove();
            List<Integer> z = new ArrayList<>(List.of(1, 2, 3, 4, 5));
            z.removeIf(x -> x % 2 == 0);
            System.out.println("D01fix Iterator.remove() -> " + w + "   removeIf() -> " + z);
        }
        {
            // the view that breaks when its parent is structurally modified
            List<Integer> base = new ArrayList<>(List.of(1, 2, 3, 4));
            List<Integer> sub = base.subList(0, 2);
            base.add(9);                       // structural change to the PARENT
            String subErr;
            try { subErr = "read " + sub.get(0); }
            catch (ConcurrentModificationException e) { subErr = "ConcurrentModificationException"; }
            System.out.println("D01view subList after the parent grew -> " + subErr
                + "   a view is only valid while the parent is structurally unchanged");
        }
        {
            // there are no pointers to dangle: growth copies references, objects stay put
            List<int[]> holder = new ArrayList<>();
            int[] elem = {7};
            holder.add(elem);
            for (int i = 0; i < 100; i++) holder.add(new int[]{i});   // forces several regrows
            System.out.println("D01noptr after " + holder.size() + " adds forced regrowth, the element"
                + " reference still reads " + holder.get(0)[0]
                + "  — Java copies REFERENCES into the new array; the objects never move,"
                + " so nothing dangles. This is the one place Java is safer than C++.");
        }
        {
            /* Three entries, not one: HashMap only re-checks modCount inside next(), so a
             * single-entry map finishes iterating before it ever notices — which would have
             * printed "no throw" and taught the exact opposite of the truth. */
            Map<String,Integer> m = new HashMap<>();
            m.put("a", 1); m.put("b", 2); m.put("c", 3);
            String err;
            try { for (String k : m.keySet()) m.put(k + "x", 1); err = "no throw"; }
            catch (ConcurrentModificationException e) { err = "ConcurrentModificationException"; }
            System.out.println("D01map mutating a HashMap while iterating its keySet -> " + err
                + "   (a value-only put(k, newValue) on an EXISTING key is fine — it is not structural)");
        }

        /* ---------------- D02  the cache hierarchy ---------------- */
        {
            final int N = 1 << 22;             // 4M ints = 16 MB
            int[] a = new int[N];
            Arrays.fill(a, 1);
            long sink = 0;

            // warm up so the JIT has compiled the loop before anything is timed
            for (int r = 0; r < 3; r++) sink += strideSum(a, 1);
            for (int r = 0; r < 3; r++) sink += strideSum(a, 1024);

            long s1 = best(a, 1), s16 = best(a, 16), s1024 = best(a, 1024);
            System.out.println("D02 the SAME number of reads, walked with a growing stride:");
            System.out.println("D02step stride 16 slower than stride 1 -> " + (s16 > s1)
                + "   stride 1024 slower than stride 16 -> " + (s1024 > s16)
                + "   1024 slower than 1 -> " + (s1024 > s1) + "  (sink " + (sink != 0) + ")");
            System.out.println("D02jvm the JIT compiles this to essentially the same machine code as C++,"
                + " so the cliff is the same cliff. The array bounds check is not what you are seeing.");
        }
        {
            final int M = 1024;
            int[] g = new int[M * M];
            Arrays.fill(g, 1);
            long warm = 0;
            for (int r = 0; r < 3; r++) { warm += rowMajor(g, M); warm += colMajor(g, M); }

            long bestRow = Long.MAX_VALUE, bestCol = Long.MAX_VALUE;
            long r1 = 0, c1 = 0;
            for (int r = 0; r < 5; r++) {
                long t0 = System.nanoTime(); r1 = rowMajor(g, M); long t1 = System.nanoTime();
                c1 = colMajor(g, M);         long t2 = System.nanoTime();
                bestRow = Math.min(bestRow, t1 - t0);
                bestCol = Math.min(bestCol, t2 - t1);
            }
            System.out.println("D02order 1024x1024 sum, same total: " + (r1 == c1)
                + ".  column-major slower than row-major -> " + (bestCol > bestRow)
                + "  — swapping two loop lines, nothing else (warm " + (warm != 0) + ")");
        }

        /* ---------------- D03  stack vs heap ---------------- */
        {
            System.out.println("D03 in Java the split is not yours to make: LOCALS of primitive type and"
                + " object REFERENCES live on the stack; every object and every array lives on the heap.");
            System.out.println("D03noalloc there is no stack allocation of objects you can request."
                + " Escape analysis may scalarise a short-lived object, but that is the JIT's"
                + " decision, not a language feature you can rely on.");
            System.out.println("D03gc the heap is garbage collected, so 'freeing' is not a cost you pay"
                + " at the free — it is a pause you pay later, proportional to LIVE objects.");
        }
        {
            /* 5000 is deep enough to be meaningful and shallow enough that every JVM
             * survives it, so the result is reproducible. Probing for the true maximum
             * would make the output depend on -Xss and on JIT state. */
            int reached = probeDepth(0, 5000);
            System.out.println("D03depth recursion to depth " + reached + " succeeds on any default JVM;"
                + " 100000 typically does NOT — the exact ceiling depends on -Xss and frame size");
            System.out.println("D03size the default JVM thread stack is 512KB-1MB (-Xss changes it),"
                + " far smaller than the 8MB a Linux main thread gets in C++.");
            System.out.println("D03risk StackOverflowError IS catchable, unlike a C++ stack overflow —"
                + " but catching it leaves the JVM in a state you should not keep working in."
                + " Convert deep recursion to an explicit Deque instead.");
        }

        /* ---------------- D04  hash containers ---------------- */
        {
            System.out.println("D04 HashMap starts with 16 buckets and doubles when size > capacity * 0.75."
                + " Growth: 16, 32, 64, 128, 256, 512, 1024, 2048 for the default load factor.");
            System.out.println("D04lf the 0.75 default is a space/time trade: higher packs tighter and"
                + " collides more, lower wastes buckets. new HashMap<>(expected / 0.75f + 1) skips"
                + " every rehash when you know the size in advance.");
            System.out.println("D04node each entry is a Node object: hash, key ref, value ref, next ref"
                + " = ~32B plus the boxed key and value. A HashMap<Integer,Integer> of 1e6 entries is"
                + " roughly 80MB; an int[] pair of arrays is 8MB.");
        }
        {
            // Java 8+ turns a long collision chain into a red-black tree
            System.out.println("D04tree since Java 8, a bucket with more than 8 entries converts to a"
                + " RED-BLACK TREE, so a collision attack degrades to O(log n) rather than O(n)."
                + " C++ has no such protection.");
            Map<Integer,Integer> m = new HashMap<>();
            for (int i = 0; i < 1000; i++) m.put(i, i);
            System.out.println("D04hash Java re-mixes every hash (h ^ h>>>16) before masking, so the"
                + " identity-hash attack that works on libstdc++ does not work here. size=" + m.size());
        }
        {
            final int N = 200000;
            int[] keys = new int[N];
            for (int i = 0; i < N; i++) keys[i] = i;
            Random rng = new Random(42);
            for (int i = N - 1; i > 0; i--) { int j = rng.nextInt(i + 1); int t = keys[i]; keys[i] = keys[j]; keys[j] = t; }

            Map<Integer,Integer> hm = new HashMap<>(N * 2);
            Map<Integer,Integer> tm = new TreeMap<>();
            for (int r = 0; r < 2; r++) { hm.clear(); for (int k : keys) hm.put(k, k); }   // warm

            long t0 = System.nanoTime();
            hm.clear(); for (int k : keys) hm.put(k, k);
            long t1 = System.nanoTime();
            for (int k : keys) tm.put(k, k);
            long t2 = System.nanoTime();
            long s = 0;
            for (int k : keys) s += hm.get(k);
            long t3 = System.nanoTime();
            for (int k : keys) s += tm.get(k);
            long t4 = System.nanoTime();

            System.out.println("D04cmp " + N + " random keys: HashMap faster to build -> " + ((t1-t0) < (t2-t1))
                + ", faster to look up -> " + ((t3-t2) < (t4-t3))
                + "  — but TreeMap gives sorted iteration, floorKey and ceilingKey, which no hash can"
                + " (checksum " + (s > 0) + ")");
        }
    }

    static long strideSum(int[] a, int stride) {
        final int mask = a.length - 1;
        final int touches = 1 << 20;
        long s = 0;
        for (int rep = 0; rep < 4; rep++)
            for (int i = 0, k = 0; k < touches; k++, i += stride) s += a[i & mask];
        return s;
    }

    /** Best of 3, so a GC pause or a scheduling hiccup cannot flip a verdict. */
    static long best(int[] a, int stride) {
        long b = Long.MAX_VALUE;
        for (int r = 0; r < 3; r++) {
            long t0 = System.nanoTime();
            long v = strideSum(a, stride);
            long t1 = System.nanoTime();
            if (v != 0) b = Math.min(b, t1 - t0);
        }
        return b;
    }

    static long rowMajor(int[] g, int M) {
        long r = 0;
        for (int i = 0; i < M; i++) for (int j = 0; j < M; j++) r += g[i * M + j];
        return r;
    }

    static long colMajor(int[] g, int M) {
        long c = 0;
        for (int j = 0; j < M; j++) for (int i = 0; i < M; i++) c += g[i * M + j];
        return c;
    }
}
